mozilla_django_oidc_db/plugins.py

```diff
@@ -49,48 +45,71 @@ def validate_settings(self) -> None:
 
-    @abstractmethod
-    def create_user(self, claims: JSONObject) -> AbstractUser | AnonymousUser:
-        """Create an authenticated user.
+        .. code:: python
 
-        This returns either a User if we need to have a new user created.
-        Otherwise, for application that don't need to create a new user whenever there is a login,
-        an AnonymousUser is returned.
-        """
-        ...
+           def view(self, request: HttpRequest) -> HttpResponse:
+               return admin_callback_view(request)
 
-    @abstractmethod
-    def update_user(self, user: AbstractUser, claims: JSONObject) -> AbstractUser:
-        """
-        Update existing user with new claims, if necessary save, and return user.
         """
         ...
 
     @abstractmethod
-    def handle_callback(self, request: HttpRequest) -> HttpResponse:
-        """Return an HttpResponse based on the callback view specified on the plugin.
+    def get_extra_params(self, request: HttpRequest, extra_params: dict) -> dict: ...
 
-        For example, for class based views:
 
-        .. code:: python
+class BaseOIDCPlugin:
+    def __init__(self, identifier: str):
+        self.identifier = identifier
 
-           def view(self, request: HttpRequest) -> HttpResponse:
-               view = self.callback_view.as_view()
-               return view(request)
+    def get_config(self) -> OIDCClient:
+        return OIDCClient.objects.resolve(self.identifier)
 
-        """
+    def get_setting(self, attr: str, *args) -> Any:
+        config = self.get_config()
+
+        return get_setting_from_config(config, attr, *args)
+
+    def get_extra_params(self, request: HttpRequest, extra_params: dict) -> dict:
+        return extra_params
+
+
+@runtime_checkable
+class AnonymousUserOIDCPluginProtocol(OIDCBasePluginProtocol, Protocol):
+    def get_or_create_user(
+        self,
+        access_token: str,
+        id_token: str,
+        payload: JSONObject,
+        request: HttpRequest,
+    ) -> AnonymousUser | None: ...
+
+
+@runtime_checkable
+class AbstractUserOIDCPluginProtocol(OIDCBasePluginProtocol, Protocol):
+    def create_user(self, claims: JSONObject) -> AbstractUser: ...
+
+    def update_user(self, user: AbstractUser, claims: JSONObject) -> AbstractUser: ...
+
+    def filter_users_by_claims(self, claims: JSONObject) -> UserManager[AbstractUser]:
+        """Return all users matching the specified subject."""
         ...
 
+    def verify_claims(self, claims: JSONObject) -> bool:
+        """Verify the provided claims to decide if authentication should be allowed."""
+        ...
+
+
+OIDCPlugin: TypeAlias = AbstractUserOIDCPluginProtocol | AnonymousUserOIDCPluginProtocol
+
 
 admin_callback_view = AdminCallbackView.as_view()
 
 
 @register(OIDC_ADMIN_CONFIG_IDENTIFIER)
-class OIDCAdminPlugin(OIDCBasePlugin):
+class OIDCAdminPlugin(BaseOIDCPlugin, AbstractUserOIDCPluginProtocol):
```

Comment:
```suggestion
class OIDCAdminPlugin(BaseOIDCPlugin):
```
do we need to inherit from the protocol? Normally you just define the protocol for the shape of the interface, but don't actually inherit from it, unless it has some default behaviour implementations. I don't know how this is with `@runtime_checkable` though, I've never used that decorator.

Reply:
I was unsure how to make it clear that this plugin implements this interface :thinking: And indeed I used later:

```
if hasattr(plugin, "get_or_create_user"):
            assert isinstance(plugin, AnonymousUserOIDCPluginProtocol)

            return plugin.get_or_create_user(
                access_token, id_token, payload, self.request
            )
```
to let the type checker know that the plugin implements `get_or_create_user`. :thinking:

Reply:
I'm not sure, let's see if @CharString can help us out?

Reply:
if it's runtime checkable, perhaps replacing the `if hasattr` with `if isinstance(plugin, AnonymousUserOIDCPluginProtocol)` is the right approach here! Then you won't need the assert anymore either.